apollo-cache-interceptor/src/commonMain/kotlin/com/apollographql/apollo/interceptor/cache/ApolloCacheInterceptor.kt

```diff
@@ -17,42 +17,86 @@ import com.apollographql.apollo.interceptor.ApolloResponse
 import com.apollographql.apollo.api.parseData
 import com.apollographql.apollo.cache.normalized.internal.normalize
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
 import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
 
 @ApolloExperimental
 class ApolloCacheInterceptor<S>(private val store: S) : ApolloRequestInterceptor where S : WriteableStore, S : ReadableStore {
 
+  private fun <D : Operation.Data> proceed(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
+    return chain.proceed(request).map {
+      if (it.response.data != null) {
+        writeToCache(request, it.response.data!!, chain.customScalarAdapters)
+      }
+      it.setFromCache(false)
+    }
+  }
+
   override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
     return flow {
-      val response = readFromCache(request)
-      if (response != null) {
-        emit(
-            ApolloResponse(
-                requestUuid = request.requestUuid,
-                response = response,
-                executionContext = request.executionContext + CacheExecutionContext(true)
-            )
-        )
-      } else {
-        chain.proceed(request).collect {
-          if (it.response.data != null) {
-            writeToCache(request, it.response.data!!)
+      val policy = request.executionContext[CacheRequestExecutionContext]?.policy ?: FetchPolicy.CACHE_FIRST
+
+      when (policy) {
+        FetchPolicy.CACHE_FIRST -> {
+          val response = readFromCache(request, chain.customScalarAdapters)
+          if (response != null) {
+            emit(response)
+          } else {
+            proceed(request, chain).collect { emit(it) }
+          }
+        }
+        FetchPolicy.NETWORK_FIRST -> {
+          proceed(request, chain)
+              .catch {
+                val response = readFromCache(request, chain.customScalarAdapters)
+                if (response != null) {
+                  emit(response)
+                }
```

Comment:
as far as I remember (might be wrong) but if we miss cache we deliver network error in old runtime implementation.

Reply:
Good catch! I fixed that. Speaking about this, I think we might want to fine tune the behaviour a bit for some fetchers:

* CACHE_ONLY -> I think this should throw on error and not return an empty response
* For all fetchers that return several values, should we introduce a "generic" exception that wraps the actual ones? It feels arbitrary to return the last one or only the network one?

Reply:
As we discussed offline, getting value from cache only in my opinion shouldn't throw error but rather return empty response. The good analogy is Map, it doesn't throw error if there is no such key. Returning empty response with a flag that it comes from cache seems like a valid case.